File: scripts/package_plugin.py
#!/usr/bin/env python3
"""Create the self-contained, source-based Paseo plugin release archive."""

from __future__ import annotations

import argparse
import json
from pathlib import Path
import tarfile


PLUGIN_FILES = (
    "client",
    "server",
    "shared",
    "index.client.tsx",
    "index.server.ts",
    "mcp.mjs",
    "paseo-plugin.json",
    "tsconfig.json",
    "README.md",
    "package.json",
    "package-lock.json",
)


def build_archive(root: Path, output_dir: Path) -> Path:
    plugin = root / "paseo-plugin"
    metadata = json.loads((plugin / "package.json").read_text(encoding="utf-8"))
    version = metadata["version"]
    output_dir.mkdir(parents=True, exist_ok=True)
    archive = output_dir / f"workspace-workbench-paseo-{version}.tar.gz"
    with tarfile.open(archive, "w:gz") as bundle:
        for entry in PLUGIN_FILES:
            source = plugin / entry
            if not source.exists():
                raise FileNotFoundError(source)
            bundle.add(source, arcname=entry, recursive=source.is_dir())
    return archive


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output-dir", type=Path, default=Path("dist"))
    args = parser.parse_args()
    root = Path(__file__).resolve().parents[1]
    archive = build_archive(root, args.output_dir)
    print(archive)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/package_plugin.mjs
#!/usr/bin/env node
/** Create the self-contained, source-based Paseo plugin archive. */

import { mkdir } from "node:fs/promises";
import { execFileSync } from "node:child_process";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { checkMetadata, readVersion } from "./version.mjs";

const root = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const entries = [
  "client", "server", "shared", "backend", "scripts", "index.client.tsx", "index.server.ts",
  "mcp.mjs", "paseo-plugin.json", "tsconfig.json", "README.md", "package.json", "package-lock.json",
];

export async function buildArchive(base = root, outputDir = join(base, "dist")) {
  const errors = await checkMetadata(base);
  if (errors.length) throw new Error(`version metadata is inconsistent:\n${errors.join("\n")}`);
  await mkdir(outputDir, { recursive: true });
  const version = await readVersion(base);
  const archive = join(resolve(outputDir), `workspace-workbench-paseo-${version}.tar.gz`);
  execFileSync("tar", ["-czf", archive, "-C", join(base, "paseo-plugin"), ...entries], { stdio: "inherit" });
  return archive;
}

if (process.argv[1] && resolve(process.argv[1]) === resolve(fileURLToPath(import.meta.url))) {
  const outputIndex = process.argv.indexOf("--output-dir");
  const outputDir = outputIndex >= 0 ? resolve(process.argv[outputIndex + 1]) : join(root, "dist");
  buildArchive(root, outputDir).then((archive) => console.log(archive)).catch((error) => { console.error(error.message); process.exitCode = 1; });
}
